sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -342,11 +343,49 @@ object FileStreamSource {
     def size: Int = map.size()
   }
 
-  private[sql] trait FileStreamSourceCleaner {
-    def clean(entry: FileEntry): Unit
+  private[sql] abstract class FileStreamSourceCleaner(
+      val fileSystem: FileSystem,
+      val sourcePath: Path) extends Logging {
+
+    private val srcPathToContainFileStreamSinkMetadata = new mutable.HashMap[Path, Boolean]
```

Comment:
This is a cache storing the result of check whether the dir contains metadata dir or not, as we may not want to do the check per batch. This is based on the assumption that a directory won't be changed from having metadata to not having metadata or vice versa, but please let me know if the assumption doesn't sound safe. I'll remove the cache and check per batch.